Initial Starlight Plugins support

HiDeoo opened the pull request:
#### What kind of changes does this PR include?

- Changes or translations of Starlight docs site content
- Changes to Starlight code

#### Description

- Implements various features of ~~#753~~#962.

This pull request is the beginning of the implementation of a Starlight Plugin API. A lot of details and discussions are available in ~~#753~~ #962.

The following features are implemented in this first PR:

- `config`
- `updateConfig()`
- `addIntegration()`
- `logger`

I tried to comment various implementation details, specially around Zod, like why the plugin API is using a totally different schema than the configuration one, why the configuration and configuration updates flowing through the plugin API are not validated (but still properly typed for user convenience) and only the final configuration is validated, etc. but if you have any questions/suggestions, please let me know.

#### Plugin Example

This PR adds a temporary `@astrojs/starlight-search-demo` plugin that demonstrates the various features of the plugin API and see how it can be used.

This example is obviously a reference to a potential `starlight-algolia` kind of plugin that has been requested multiple times on Discord and would also be needed for the Astro docs site. The idea was to see the code required to implement such a plugin and having a custom search bar by only adding the following lines to the `astro.config.mjs` file:

```js
import starlightSearchDemo from "@astrojs/starlight-search-demo";

starlight({
  plugins: [starlightSearchDemo({ apiKey: "123456" })],
});
```

Note that this plugin is only meant for demo purposes. It does nothing and even tho we briefly discussed the idea of having official plugins and if we ever decide to make an official Algolia plugin, it feels like this should be done in a separate PR or maybe even a separate repository with all expected features from such a plugin (proper CSS styling, pagefind disabled, tests, etc.).

_Note: I think my personal important takeaway from this example is that the `injectSettings()` API discussed in #753 (and potentially in #861 too for some API changes) could be really useful._

#### Documentation

This PR adds a "Plugins Reference" page inspired by the [Astro Integration API](https://docs.astro.build/en/reference/integrations-reference/) documentation page.
I'm not sure if a guide is needed for the plugin API at this point, as the "use a plugin" section would be pretty much a repeat of the `plugins` section of the Configuration Reference and may change between plugins and a "create a plugin" section, which I think does not exist for the Astro Integration API, would be pretty much merging the various examples from the "Plugins Reference" page.

I also suggested a change to the showcase to add a new section dedicated for plugins so it can be linked from various places. Obviously, we would need to wait for the first plugin to be published before doing that if we decide to go that way.

#### Remaining tasks

These tasks should be completed before merging this PR:

- [x] Remove the Starlight Search Demo plugin in its current state
- [x] Remove the Starlight Search Demo plugin from the docs configuration file
- [x] Remove all `TODO(HiDeoo)` references.
- [x] Add a changeset.

delucis commented:
Super exciting to see this! Will make sure to get you a thorough review ASAP.

delucis commented on packages/starlight/utils/plugins.ts:
```diff
@@ -0,0 +1,143 @@
+	let userConfig = config;
+	const integrations: AstroIntegration[] = [];
+
+	for (const { name, plugin } of parsedPluginsConfig.data.plugins) {
```
Should we support processing plugins added by preceding plugins? If so, maybe we need something like this?

```suggestion
	for (let i = 0; i < userConfig.plugins.length; i++) {
		const { name, plugin } = userConfig.plugins[i];
```

HiDeoo replied:
I thought about this during the initial prototype and decided to leave it out for now. My reasoning (and maybe wrong) was that it would maybe be more "natural"/explicit to add a `addPlugin` like `addIntegration` considering that we only want to support adding plugins but not remove/re-order them while running them which I think could be less obvious with a helper called `updateConfig`.

If we decide to allow this in `updateConfig`, we would indeed need to tweak the syntax like you suggested and also extract the updated plugins config from the updated Starlight config and also re-run the plugin validation obviously.

delucis replied:
That does make sense to me.

I will share some thoughts from the discussion of when we enabled integrations to add integrations in Astro’s `updateConfig()`. Initially I made an `addIntegration()` helper, but Blu pointed out that ideally `updateConfig()` works for all config options so a user doesn’t have to know what they can and can’t use that hook for (otherwise you risk proliferating `updateSidebar()`, `updateHead()` etc. for any more complex option).

I think the issue in Starlight’s case is that Astro’s [`mergeConfig()`](https://github.com/withastro/astro/blob/262cef2487c7494bd8d23b4ab27bfcdf1870a111/packages/astro/src/core/config/merge.ts) is generally able to not worry about order, so merging `{ option: [something] }` and `updateConfig({ option: [somethingElse] })` as `{ option: [something, somethingElse] }` is safe (and also nicer for the dev to not worry about spreading all the right things). But in Starlight I can totally imagine wanting to do stuff like prepend or filter sidebar items etc., which the `mergeConfig()` approach prevents. So I think I agree with forcing plugin developers to merge config manually even if this does place some trust in them to not break user config.

`plugins` is the only problematic field here right? I wonder if we should error if a plugin tries to mutate it?

HiDeoo replied:
> … Blu pointed out that ideally `updateConfig()` works for all config options so a user doesn’t have to know what they can and can’t use that hook for …

This is a very good point. 

> I think the issue in Starlight’s case is that Astro’s `mergeConfig()` is generally able to not worry about order

My initial implementation actually had a `mergeConfig()` which was a slimmed down version of Astro’s `mergeConfig()` but as you mentioned, this is not ideal in the case a Starlight.

> `plugins` is the only problematic field here right? 

This is indeed the only one.

> I wonder if we should error if a plugin tries to mutate it?

This would already happen when validating the config after the plugins have run but maybe you mean adding an explicit check in the `updateConfig()` hook?

delucis replied:
> This would already happen when validating the config after the plugins have run but maybe you mean adding an explicit check in the `updateConfig()` hook?

I think you answered this in https://github.com/withastro/starlight/pull/942#discussion_r1378878069 but validating in `updateConfig()` or after would both be fine. I guess inside `updateConfig()` we might have the advantage of a nice error message that says which plugin was responsible.

delucis commented on packages/starlight/utils/plugins.ts:
```diff
@@ -0,0 +1,143 @@
+	hooks: z.object({}).passthrough().default({}),
+}) as z.Schema<AstroIntegration>;
+
+/**
+ * A plugin `config` and `updateConfig` argument are purposely not validated using the Starlight
+ * user config schema but properly typed for user convenience because we do not want to run any of
+ * the Zod `transform`s used in the user config schema when running plugins.
+ * The final user config should be validated by the caller after running all the plugins.
+ */
+const starlightPluginSchema = z.object({
```
I’m not sure I follow why we can’t include `plugins` in the main user config schema? Would you mind explaining a bit more?

HiDeoo replied:
Is this question related to the highlighted comment or more of a general question?

If this is directly relateed to the comment, I think my comment is not obvious enough. I try to explain why we are not using something like:

```js
updateConfig: z.function(
  z.tuple([UserConfigSchema]),
  //          ^ this is the problem described in the comment
  z.void()
)
```

If this is not directly to the comment, this would be technically possible to add the plugins to the user config schema. The reason I did not do that are:

- There is __at the moment__ in the code only 1 place (the Starlight integration input) where we need a type that contains a config with the plugins. 
- Past that point, we only need the config without the plugins (when passing it down to plugins, when a plugin updates the config, for passing it to the Vite plugin, etc.)
- I did not want to re-validate the plugins config when re-validating the user config after the plugins have run.
- I thought it would be easier to use a TypeScript union once vs omitting the plugin from the user config schema with the Zod API to get the same result.

This could definitely be revisited if you think it would be better to have the plugins in the user config schema.

delucis replied:
> Past that point, we only need the config without the plugins (when passing it down to plugins, when a plugin updates the config, for passing it to the Vite plugin, etc.)

Wouldn’t other plugins want to know what plugins they are configured alongside potentially?

Makes sense for the Vite plugin though, where functions would be stripped during serialization.

HiDeoo replied:
> Wouldn’t other plugins want to know what plugins they are configured alongside potentially?

I didn't think of this but it's a very good point.

So I guess a potential solution would be:

- Add `plugins` to the user config schema and refactor the various types/calls to follow this change
- Pass down the `plugins` too to the provided `config` to a plugin
- As per your comment above, check if `plugins` is being mutated in the `updateConfig()` hook and error if so

Is this something that looks good to you?

delucis replied:
Yeah, sounds good to me!

delucis reviewed (commented):
Super exciting work @HiDeoo! Left a few initial thoughts and questions.

HiDeoo commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -0,0 +1,99 @@
+## Quick API Reference
+
+```ts
+interface StarlightPlugin {
+  name: string;
+  plugin: (options: {
+    config: StarlightUserConfig;
+    updateConfig: (newConfig: StarlightUserConfig) => void;
+    addIntegration: (integration: AstroIntegration) => void;
+    logger: AstroIntegrationLogger;
+  }) => void | Promise<void>;
+}
````
I was never a huge fan of the `plugin()` method name but did not managed at the time to come up with a better name I liked. I personally like `setup()` so I'm all for it.

Regarding the structure, I personally do not have a strong preference, so I think if you have one, we should go with it otherwise I guess mirroring Astro's structure may be the safest bet?

delucis replied:
No strong preference from me, so yeah, +1 to matching Astro’s shape 👍

HiDeoo commented on packages/starlight/index.ts:
```diff
@@ -7,25 +7,29 @@ import { starlightAsides } from './integrations/asides';
+		name: '@astrojs/starlight',
+		hooks: {
+			'astro:config:setup': async ({ config, injectRoute, logger, updateConfig }) => {
+				const { integrations, userConfig } = await runPlugins(opts, logger);
+				const parsedConfig = StarlightConfigSchema.safeParse(userConfig, { errorMap });
```
> Would be nice to add a comment regarding the order here.

Sure, I'll try to describe it better.

> I’m also wondering if we want to parse twice so we can error nicely before plugins run. I’m guessing currently we’d have the risk that a user provides bad config that causes some error in a plugin that Zod could catch but the user would see an error thrown by the plugin?

We definitely could extract the validation part in a function and run it before and after and it could indeed improve the experience when dealing with an error.

delucis commented on packages/starlight/utils/plugins.ts:
```diff
@@ -0,0 +1,143 @@
+	 * Plugin function called with an object containing various values that can be used by the
+	 * plugin to interact with Starlight.
+	 */
+	plugin: z.function(
```
Would it make sense to also pass some [data from the `astro:config:setup`](https://docs.astro.build/en/reference/integrations-reference/#astroconfigsetup) hook here? I think it’s OK to require adding an integration to use any of the functions, but it would probably be helpful for plugins to access [`config`](https://docs.astro.build/en/reference/integrations-reference/#config-option), [`command`](https://docs.astro.build/en/reference/integrations-reference/#command-option), and [`isRestart`](https://docs.astro.build/en/reference/integrations-reference/#isrestart-option). 

Example use cases:
- check if an integration is already in the Astro config before adding it
- only run some initialisation code on dev/build start, but not when the dev server restarts

delucis replied:
Yeah, because otherwise you could be in the weird situation of using `addIntegration()` to add an integration that reads `config` and _then_ itself adds an integration based on that, which is pretty clunky. (The use case I have in mind is the Algolia plugin, which could add Preact or React depending on what a user is already using.)

HiDeoo commented:
Thanks a lot for the initial feedback and discussions.

I think all of the points raised are now actionable and I can get back to it and prepare an updated version including all the suggested changes 👍

HiDeoo commented:
> Add plugins to the user config schema

Turns out this one is a little bit "annoying" due to the recursive types (a user config including plugins which contains a function that pass down a user config that contains plugins…) and our tranforms.

I've setup a small [repro](https://tsplay.dev/WPMPZw) in the TS playground of the issue. I've found so far 1 workaround that I describe in the playgraound but imo this is pretty ugly and I'm not sure if it's even worth it.

I'll continue to explore this as I may be missing something obvious or a better way to do this but so far, no luck.

HiDeoo commented:
Just pushed a few changes relative to the first review:

- Updated the shape of a plugin to match the one from an Astro integration.
- The configuration passed down to a plugin now also contains the `plugins` list.
- Refactor the entire validation process:
  1. Before running plugins, the user-provided configuration is validated (we can provided an early feedback to the user)
  2. Then, the user-provided plugins are validated (we can provided an early feedback to the user and indicate that a plugin is not valid)
  3. The plugins are executed
     3.1. If a plugin updates the configuration, we check that it does not try to mutate the `plugins` list (we have a dedicated error message indicating which plugin is responsible for the error)
     3.2. If a plugin updates the configuration, we validate the new configuration (if the merged configuration is not valid, we have a dedicated error message indicating which plugin is responsible for the error)
  4. We return the final validated configuration and the list of Astro integrations to add
- I added a few comments to help understand the flow of the code regarding the validation process
- A plugin now receives a few new options (`astroConfig`, `command`, and `isRestart`)
- I updated the tests and documentation to reflect the changes
- I slightly updated the demo
  - Fix the color contrast issue ^^
  - I added a button using Preact to show how a plugin should properly test if an Astro integration is already present before adding it

If I did not miss anything, this should covers all the feedback from the first review except the one about adding `plugins` directly to the user config schema. As commented right above, I still have not a found a proper way to do this that works as expected in all cases but I will keep looking into it.

delucis commented on packages/starlight/utils/plugins.ts:
```diff
@@ -0,0 +1,223 @@
+				// Collect any Astro integrations added by the plugin.
+				integrations.push(integration);
+			},
+			astroConfig: context.config,
```
I think as a plugin author I’d expect `astroConfig.integrations` here to include the updated values from previous plugins (or at least to have another way to check this, but here probably make most sense?)

Thoughts on doing something like this? On the one hand it’s slightly lying, but on the other hand, in practice this is what is happening or “will” happen once we update the Astro config after plugins run?

```suggestion
			astroConfig: { ...context.config, integrations: [...context.config.integration, ...integrations] },
```

Effectively mirrors what we do with the `updateConfig()` hook where each plugin is passed `userConfig` as mutated by the previous.

HiDeoo replied:
Good catch, and I think this is a good solution you’ve come up with. I implemented this and added a test for it.

delucis reviewed (commented):
Took the liberty of resolving conflicts and updating this branch @HiDeoo! Here are a few comments on the latest version — I’ve looked at everything except docs so far.

What’s the state of the typing issue? Is it kind of resolved in the current code or still needs a solution?

HiDeoo commented on packages/starlight/index.ts:
```diff
@@ -37,34 +44,25 @@ export default function StarlightIntegration(opts: StarlightUserConfig): AstroIn
 					entryPoint: '@astrojs/starlight/index.astro',
 				});
-				const integrations: AstroIntegration[] = [];
 				if (!config.integrations.find(({ name }) => name === 'astro-expressive-code')) {
```
Really good catch, and I even think this should actually check both 🤔 I updated the PR to do so.

HiDeoo commented on packages/starlight/__tests__/test-config.ts:
```diff
@@ -18,14 +21,29 @@ export function defineVitestConfig(
+	warn = () => undefined;
+	error = () => undefined;
+	debug = () => undefined;
+}
```
This was actually where I started writing this code but unfortunately it seems to not be possible as `test-utils` cannot be imported in `test-config` because it imports `vitest` and would give the following error:

```
Error: Vitest failed to access its internal state.
```

delucis replied:
Ahhhh, got it. Wondered if there might be something like that. OK then I guess we can leave it like this for now. Other option would be a separate file like `test-plugin-utils` or something, but probably overkill for so little code.

HiDeoo commented on docs/src/content/docs/showcase.mdx:
Definitely, I thought I mentioned that in the PR description in the documentation section but I might have forgotten to do so. This change was mostly a suggestion to see what we could do on that side of things and see if it could somehow bring ideas to the table.

This would need to be removed until we get published plugins (it's in the list of remaining tasks at the end of the PR description).

delucis commented:
> I was continuing my work locally and did not care about updating it as it was not relevant to my current work.

Ooh, what are you planning? Was wondering if this might be getting towards the point where we could think about releasing it? What would you still like to do before it’s ready for a final review?

HiDeoo commented:
> Ooh, what are you planning? Was wondering if this might be getting towards the point where we could think about releasing it? What would you still like to do before it’s ready for a final review?

Oh sorry, I was not clear enough, I've been solely working on trying to make that refactor of the user-config schema working with `plugins` being part it. Nothing else.

Feature-wise, I think we are good compared to the initial plan (with potential future improvements as we discussed in the "RFC" issue) and that's also why I suggested in my last point that going with the current implementation without this refactor is a viable option _imo_.

delucis commented:
OK, great! Then let’s mark this as ready for review and move onto polishing it up. I’ll take a look at the docs in a bit 🙌

delucis reviewed (commented):
Here’s the docs review as promised! 🙌

HiDeoo commented:
- Went ahead and added the `test-plugin-utils` you suggested as I think it makes more sense.
- Added the suggested documentation changes.

A few notes:

- There is still an "easy" reference in `reference/plugins.md`. I think the context is a bit different from the one removed but I prefer to let you know and confirm if it's ok or not in case this was missed.
- I tried to rephrase the `updateConfig` sentence but I'm sure it's still not clear enough so feel free to ditch it.
- I updated the EC highlight for the `addIntegration` example which I think was wrong.

Thanks for the review and especially the comments, I hope to get better at this with time by understanding the reasoning behind the changes. Super helpful! :raised_hands:

delucis commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -0,0 +1,152 @@
+
+For example, to add a new [`social`](/reference/configuration/#social) media account to the configuration without overriding the existing ones:
+
+```ts {5-10}
+// plugin.ts
+export default {
+  name: 'add-twitter-plugin',
+  plugin({ config, updateConfig }) {
+    updateConfig({
+      social: {
+        ...config.social,
+        twitter: 'astrodotbuild',
+      },
+    });
+  },
````
Totally missed this on my first read through, but just realising the examples all still use the old shape 😅
````suggestion
```ts {6-11}
// plugin.ts
export default {
  name: 'add-twitter-plugin',
  hooks: {
    setup({ config, updateConfig }) {
      updateConfig({
        social: {
          ...config.social,
          twitter: 'astrodotbuild',
        },
      });
    },
  },
````

HiDeoo replied:
Damn, good catch, can't believe I missed that so many times too 😨

delucis commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -0,0 +1,152 @@
+A callback function to add an [Astro integration](https://docs.astro.build/en/reference/integrations-reference/) required by the plugin.
+For example, to add the [React Astro integration](https://docs.astro.build/en/guides/integrations-guide/react/):
+
+```ts {13}
+// plugin.ts
+import react from '@astrojs/react';
+
+export default {
+  name: 'plugin-using-react',
+  plugin({ addIntegration, astroConfig }) {
+    const isReactLoaded = astroConfig.integrations.find(
+      ({ name }) => name === '@astrojs/react'
+    );
+
+    // Only add the React integration if it's not already loaded.
+    if (!isReactLoaded) {
+      addIntegration(react());
+    }
+  },
````
````suggestion
```ts {14}
// plugin.ts
import react from '@astrojs/react';

export default {
  name: 'plugin-using-react',
  hooks: {
    plugin({ addIntegration, astroConfig }) {
      const isReactLoaded = astroConfig.integrations.find(
        ({ name }) => name === '@astrojs/react'
      );

      // Only add the React integration if it's not already loaded.
      if (!isReactLoaded) {
        addIntegration(react());
      }
    },
  },
````

delucis commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -0,0 +1,152 @@
+An instance of the [Astro integration logger](https://docs.astro.build/en/reference/integrations-reference/#astrointegrationlogger) that you can use to write logs.
+All logged messages will be prefixed with the plugin name.
+
+```ts {5}
+// plugin.ts
+export default {
+  name: 'long-process-plugin',
+  plugin({ logger }) {
+    logger.info('Starting long process…');
+    // Some long process…
+  },
````
````suggestion
```ts {6}
// plugin.ts
export default {
  name: 'long-process-plugin',
  hooks: {
    plugin({ logger }) {
      logger.info('Starting long process…');
      // Some long process…
    },
  },
````

delucis commented:
@HiDeoo do you want to take care of cleaning up the outstanding `TODO`s or would you like a hand getting this ready for release?

HiDeoo commented:
> @HiDeoo do you want to take care of cleaning up the outstanding `TODO`s or would you like a hand getting this ready for release?

If you think it's ok, I can indeed get started on cleaning the PR when I get the time.

The changeset is probably the one thing where I might need help/review but I'll push one and we can see how it goes.

delucis commented:
Sounds good, thank you! Yes, we can do a final docs and changeset review once those are in.

HiDeoo commented:
Updated the PR with the following changes:

- Remove demo plugin and its usage in the docs
- Revert the showcase page to a pre-plugin version
- Remove plugin showcase references from `reference/configuration.mdx` & `reference/plugins.md`
- Search and remove all `TODO(HiDeoo)` references
- Add changeset (not sure what level of detail is expected here)

delucis commented on docs/src/content/docs/reference/plugins.md:
```diff
@@ -0,0 +1,158 @@
+**type:** `(integration: AstroIntegration) => void`
+
+A callback function to add an [Astro integration](https://docs.astro.build/en/reference/integrations-reference/) required by the plugin.
+For example, to add the [React Astro integration](https://docs.astro.build/en/guides/integrations-guide/react/):
```
Small suggestion to add the context around first checking before registering a new integration (which I think is a good practice for most integrations):

```suggestion

In the following example, the plugin first checks if [Astro’s React integration](https://docs.astro.build/en/guides/integrations-guide/react/) is configured and, if it isn’t, uses `addIntegration()` to add it:
```

delucis commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -0,0 +1,158 @@
+A callback function to add an [Astro integration](https://docs.astro.build/en/reference/integrations-reference/) required by the plugin.
+For example, to add the [React Astro integration](https://docs.astro.build/en/guides/integrations-guide/react/):
+
+```ts {14}
````
Maybe makes sense to also highlight the `addIntegration` in the plugin options?

````suggestion
```ts {14} "addIntegration,"
````

delucis commented on docs/src/content/docs/reference/plugins.md:
```diff
@@ -0,0 +1,158 @@
+**type:** `AstroConfig`
+
+A read-only copy of the user-supplied [Astro configuration](https://docs.astro.build/en/reference/configuration-reference/).
+This configuration is resolved before any other integrations have run.
```
Is this 100% accurate? I’m pretty sure Starlight receives the config as transformed by any preceding integrations. If it _is_ accurate, maybe should be “Astro integrations” for clarity.

HiDeoo replied:
Good question, this is totally based on https://docs.astro.build/en/reference/integrations-reference/#config-option :thinking:

HiDeoo replied:
Should we just remove the sentence then?

PS: not sure why everything is red now, looks like it's related to the size check, I'll try to see why 🤔

delucis replied:
Yes, I think removing the sentence is fine.

Yeah, not sure why either — looks like the docs site is failing to build all of a sudden.

delucis reviewed (commented):
Really excited we’re getting so close!

I’ve left a last pass of comments on the docs & changeset, but then I think we should be good to go. 🚀

delucis commented:
Huh, not sure why CI started failing 🤔

delucis commented:
The CI failure is very odd — just ran the build with no issues locally.

HiDeoo commented:
> The CI failure is very odd — just ran the build with no issues locally.

Same, working fine locally 🤔

HiDeoo commented:
Yeah I'm a bit confused as there is no `userConfig` anymore in the `index.ts` file anymore

delucis commented:
Merged and fixed that to see if it helps!

delucis commented:
Looks like that worked? I guess somehow there was some weird code in the CI job?

delucis reviewed (approved):
STAMP OF APPROVAL! ✅

HiDeoo commented:
> Looks like that worked? I guess somehow there was some weird code in the CI job?

Yeah, I was afraid to try this, see it works and have no clue on why 🤣 I do not have a better explanation ^^

delucis commented:
Well, here goes nothing… 🤞
